internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,373 @@
+
+	b, _ := json.MarshalIndent(receipt, "", "  ")
+	l.store.Put(lookupKey, b)
+
+	// build the index for "from"
+	fromKey := fmt.Sprintf("from:%s:%s", (*receipt)["from"], lookupKey)
+	l.store.Put(fromKey, []byte(lookupKey))
+
+	// build the index for "to" if a value is present
+	to := (*receipt)["to"]
+	if to != "" {
+		toKey := fmt.Sprintf("to:%s:%s", to, lookupKey)
+		l.store.Put(toKey, []byte(lookupKey))
+	}
+
+	// build the indiex for "receivedAt"
+	receivedAtKey := fmt.Sprintf("receivedAt:%d:%s", (*receipt)["receivedAt"], lookupKey)
+	l.store.Put(receivedAtKey, []byte(lookupKey))
+
+	// insert the lookup entry for GetReceipt()
+	return l.store.Put(requestID, []byte(lookupKey))
+}
+
+// GetReceipts Returns recent receipts with skip, limit and other query parameters
+func (l *levelDBReceipts) GetReceipts(skip, limit int, ids []string, sinceEpochMS int64, from, to, start string) (*[]map[string]interface{}, error) {
+	// the application of the parameters are implemented to match mongo queries:
+	// - find the starting point:
+	//   - if "start" is present, use it
+	//   - otherwise use "Last()"
+	// - if "skip" is present, forward to the count
+	// - if "ids" are present, use them to look up the specific entries and filter out the entries falling out of the cursor range
+	// - if "from" or "to" are present, look up the entries using the "from:[address]" and "to:[address]" prefix then work out the intersection of the [lookupKey] segments
+	var itr kvstore.KVIterator
+	var endKey string
+	results := []map[string]interface{}{}
+	if sinceEpochMS > 0 {
+		// locate the iterator range limit
+		keyRange := l.findEndPoint(sinceEpochMS)
+		if keyRange != nil {
+			itr = l.store.NewIterator(keyRange)
+			endKey = string(string(keyRange.Limit))
+		} else {
+			// no entries match the sinceEpochMS, return empty
+			return &results, nil
+		}
+	} else {
+		// create the iterator without a range
+		itr = l.store.NewIterator()
+	}
+	defer itr.Release()
+
+	if limit == 0 {
+		limit = l.defaultLimit
+	}
+
+	// if we have reference points, like list of IDs and "from/to" parameters, process using the reference points
+	var lookupKeys []string
+	var lookupKeysByIDs []string
+	if len(ids) > 0 {
+		// use the list of IDs to search for the result entries
+		lookupKeysByIDs = l.getLookupKeysByIDs(ids, start, endKey)
+	}
+	var lookupKeysByFromAndTo []string
+	if from != "" || to != "" {
+		lookupKeysByFromAndTo = l.getLookupKeysByFromAndTo(from, to, start, endKey, limit)
+	}
+	if len(ids) > 0 && from == "" && to == "" {
+		lookupKeys = lookupKeysByIDs
+	} else if len(ids) == 0 && (from != "" || to != "") {
+		lookupKeys = lookupKeysByFromAndTo
+	} else if len(ids) > 0 && (from != "" || to != "") {
+		lookupKeys = intersect(lookupKeysByIDs, lookupKeysByFromAndTo)
+		sort.Sort(sort.Reverse(sort.StringSlice(lookupKeys)))
+		results := l.getReceiptsByLookupKey(lookupKeys, limit)
+		return results, nil
+	}
+
+	// no reference points, iterate normally
+	index := 0
```

Comment:
It might be more readable to have this as a separate `getReceiptsNoFilter` function that's called.

... I'll admit it does seem duplicative to have this, as I guess the only difference here is there's no filtering on the loop.
But I understand the loop with filtering is rather complex.